### PR TITLE
Tiptap RTE: Fix style menu class detection on marks and stuck toggles

### PR DESCRIPTION
**Pull request by leekelleher**
## Description

Following on from PR #23891.

Style menu items configured with only a `class` (no tag) previously only checked ancestor node types for a match, so they never showed as active/selected when the class actually lived on a mark like bold or a link, or on a selected image. They also missed the case where the class sits on an outer element like a blockquote wrapping a plain paragraph.

That second case had a worse symptom: toggling such an item did not just fail to detect it, it applied per element type independently, so removing the class from the blockquote added it right back to the paragraph inside it. The button stayed highlighted, and clicking it again just swapped which element carried the class, forever.

Both are fixed now: detection also checks marks and node selections, and toggling makes one add/remove decision for the whole selection instead of one per element type. While in there, the whole-class-name comparison (so `size-1` doesn't match `size-10`) is shared between the toggle command and the style menu instead of living as two copies that could drift apart.

### How to test?

1. In `src/Umbraco.Web.UI.Client/src/packages/tiptap/extensions/style-select/manifests.ts`, temporarily add a class-only item, e.g. `{ label: 'Callout', data: { class: 'callout' } }`, to the `Containers` group.
2. Run the backoffice locally and open a Rich Text Editor property whose toolbar includes **Style Select**.
3. Select some text, apply **Bold**, then click **Style Select → Callout**. The item now highlights while the caret stays in the bold text (previously it never did).
4. Put the caret in a blockquote wrapping a plain paragraph, apply **Callout**, then click it again. It clears fully and the item un-highlights (previously it stayed highlighted, alternating which element held the class).
5. Revert the manifest edit from step 1.

### CODE REVIEW

File: src/Umbraco.Web.UI.Client/src/packages/tiptap/extensions/style-menu/style-menu.tiptap-toolbar-api.ts [hunk 1]
```diff
@@ -54,33 +56,42 @@ export default class UmbTiptapToolbarStyleMenuApi extends UmbTiptapToolbarElemen
...
 		for (let depth = $from.depth; depth > 0; depth--) {
 			if (this.#hasAttributes($from.node(depth).attrs, id, className)) return true;
 		}
 		return false;
 	}
 
+	#hasMarkWithAttributes(editor: Editor, id?: string, className?: string): boolean {
+		const { state } = editor;
+		const marks = state.storedMarks ?? state.selection.$from.marks();
+		return marks.some((mark) => this.#hasAttributes(mark.attrs, id, className));
```
[the remark this answers is not in the document]
Comment (leekelleher) [reply src/Umbraco.Web.UI.Client/src/packages/tiptap/extensions/style-menu/style-menu.tiptap-toolbar-api.ts]: Fixed in 987a93b: `#hasMarkWithAttributes` now mirrors the range Tiptap's own `getMarkAttributes` uses for a mark command — `storedMarks ?? $from.marks()` for a collapsed selection, `doc.nodesBetween(from, to)` otherwise — so detection and `execute`'s toggle agree on the same range. Added a regression test with the caret spanning a selection where the class-bearing mark starts partway through it.

**Comment by leekelleher** [comment]
Pushed a follow-up on review feedback:

- Fixed the mark-detection scope gap Copilot flagged on `#hasMarkWithAttributes` (resolved above) — `987a93b`.
- Fixed the two SonarQube issues on `class-names.function.ts` (`String(value ?? '')` stringifying non-strings, and the `filter` arrow being equivalent to `Boolean`) — `b7c18762cb2`.
- Added the toggle-off symmetry test the Claude review suggested — `999d41b48cf`.

Declining two items, leaving them for a follow-up rather than widening this bugfix:

- Copilot's suppressed comment on `html-attr-class.tiptap-extension.ts` (`editor.getAttributes(t)` reading only one node's attrs when a selection spans multiple nodes of the same type). That's pre-existing — this PR only hoisted the decision out of the per-type `.map()`, the same single-node read already fed the `updateAttributes`/`resetAttributes` calls below it. Fixing it properly means a per-node traversal rewrite of `toggleClassName`, not a small addition.
- Claude's suggestion to document that `hasClassNames(x, '')` returns `true`. That case is already pinned by a named test (`class-names.function.test.ts` — "is true when no class names are requested"), and no caller can reach it since every call site guards on `!className` first.

CodeScene also flagged Primitive Obsession / String Heavy Function Arguments on `style-menu.tiptap-toolbar-api.ts` — addressed as part of `987a93b` by passing the style item's own `data` object through the private helpers instead of two loose `id?/className?` strings.

**Review by AndyButland — CHANGES_REQUESTED** [review]
Looks good @leekelleher, though I had to adjust a bit the test steps to get cases where I could replicate the issue on `v17/dev` and verify the fix on this PR's branch.  I could do that though, so this resolves the expected issues.  I found a couple of edge cases though.

What follows replays the testing I've done so I can be replicated, detailing the issues proved fixed and how to see, and perhaps fix, the edge cases found.

---

## Test steps 

Tests 1–2 reproduce the bugs the PR fixes, so they need **unfixed `v17/dev`**. Tests 3–5 reproduce issues remaining in the PR itself, so they need the **PR branch** `v17/bugfix/tiptap-style-menu-class-attribute-matching`. 

## Part 1 — the bugs this PR fixes (run on `v17/dev`)

### Test 1 — stuck toggle: the class moves instead of clearing

1. Type `text`, leave the caret in it.
2. Style Select → Containers → **Callout**. Source: `<p class="callout">text</p>`
3. Style Select → Containers → **Block quote**. Source: `<blockquote><p class="callout">text</p></blockquote>` ← asymmetric state reached
4. Open Style Select → Containers. **Callout is highlighted.** Click it.
5. **Bug:** source is now `<blockquote class="callout"><p>text</p></blockquote>` — the class moved to the blockquote rather than clearing. Reopen the menu: Callout is still highlighted.
6. Click it again → back to `<blockquote><p class="callout">text</p></blockquote>`. It alternates between the two elements indefinitely and never turns off.

**Expected and found with this PR:** step 5 removes the class from both elements (`<blockquote><p>text</p></blockquote>`) and the item un-highlights.

### Test 2 — a class on a mark is not detected

1. Type `some text here`.
2. Select the word `text` → **Bold**.
3. With `text` still selected: Style Select → Containers → **Callout**. Source: `<p class="callout">some <strong class="callout">text</strong> here</p>` — the class lands on the paragraph *and* the bold mark, which is expected and unchanged by the PR; a class-only item writes to every configured type in the selection. This is why step 4 is needed to reach a state where only the mark carries it.
4. Select only the word `here` (so the selection contains no bold) → Style Select → **Callout**. Source: `<p>some <strong class="callout">text</strong> here</p>` ← class now on the `<strong>` only
5. **Bug:** put the caret inside the bold word `text` and open Style Select → Containers. Callout is **not** highlighted, even though the class is on the mark around the caret.
6. Click Callout anyway: the class jumps to the paragraph (`<p class="callout">some <strong>text</strong> here</p>`) and only *then* does the item highlight.

**Expected and found with this PR:** step 5 shows Callout highlighted; step 6 clears the class.

## Part 2 — issues remaining in the PR (run on the PR branch)

All three share one root cause: `isActive` reads a **narrower** slice of the document than `execute` does. `execute` toggles through `editor.getAttributes(type)`, which scans the whole selection range; the new detection helpers scan less than that. Wherever they disagree the item reads *not highlighted* while clicking it performs a *removal* — the inverse of the Part 1 symptom, and arguably worse, because an un-highlighted item promises to add the class and instead strips it.

| Test | Where detection looks | Where `execute` looks | Measured result |
| --- | --- | --- | --- |
| 3 | `storedMarks` only, whenever it is set | `storedMarks` **plus** the marks at the caret | `isActive` false, `getAttributes('bold')` = `{class: 'callout'}` |
| 4 | ancestors of the selection **start** | every node in the range, last match wins | `isActive` false, `getAttributes('paragraph')` = `{class: 'callout'}` |
| 5 | the selected node's own attributes | the selected node **and its descendants** | `isActive` false, `getAttributes('image')` = `{class: 'callout', …}` |

### Test 3 — pressing Ctrl+B at the caret hides a class that is still there

`#hasMarkWithAttributes` uses `state.storedMarks ?? $from.marks()`, but tiptap's `getMarkAttributes` uses `storedMarks` **and** `$head.marks()`. Any mark toggle at a collapsed caret sets `storedMarks` to a non-null array, at which point `??` discards the marks actually present in the document.

1. Reach the mark-only state using Test 2 steps 1–4 — those steps behave the same with the PR in place, ending at `<p>some <strong class="callout">text</strong> here</p>`.
2. Put the caret inside the bold word `text`, open Style Select → Containers: **Callout is highlighted** (this is the PR's fix working).
3. Close the menu. With the caret still inside the bold word, press **Ctrl+B** — the normal "stop bolding from here" gesture. Nothing visibly changes and the source is untouched: the `<strong class="callout">` is still there.
4. **Bug:** reopen Style Select → Containers. Callout is **no longer highlighted**.
5. Press an arrow key to move the caret one character, then reopen the menu: highlighted again. Nothing about the document changed between steps 3 and 5 — only the stored-mark state — which is the tell.
6. To see the consequence, repeat step 3 and then click the un-highlighted Callout: the class is **removed** (`<p>some <strong>text</strong> here</p>`), i.e. the item that read "off" also acted "off".

**Suggested fix:** `const marks = [...(state.storedMarks ?? []), ...state.selection.$head.marks()];`

### Test 4 — a class on a later block of the selection is not detected

`#hasAncestorNodeWithAttributes` walks the ancestors of `$from` only, while `execute` reads node attributes through `getNodeAttributes`, which walks `nodesBetween(from, to)` and takes the **last** match. So the two disagree whenever the class sits on a block after the one containing the start of the selection. `$from` is always the earlier document position, so drag direction is irrelevant.

1. Type `first`, press Enter, type `second`.
2. Put the caret in the **second** paragraph → Style Select → Containers → **Callout**. Source: `<p>first</p><p class="callout">second</p>`
3. Select across both paragraphs — from inside `first` through into `second` (drag, or Shift+Down).
4. **Bug:** open Style Select → Containers. Callout is **not** highlighted, although the second paragraph in the selection carries it.
5. Click Callout, expecting it applied to the whole selection. Instead the class is **removed**: `<p>first</p><p>second</p>`.
6. Contrast, to confirm it is position-dependent: put the class on the **first** paragraph instead (`<p class="callout">first</p><p>second</p>`), select across both again, and the item **is** highlighted.

**Suggested fix:** extend the `nodesBetween` walk in `#hasMarkWithAttributes` to test `node.attrs` as well as `node.marks`, keeping the ancestor walk for the collapsed case.

### Test 5 — a class on an image inside a selected figure is not detected

`#hasSelectedNodeWithAttributes` tests only the selected node's own attributes, but `getAttributes('image')` descends into it. A `figure` is `atom: true, selectable: true`, so clicking an image with a caption selects the figure and the inner `` is never examined. Like the mark-only case, the menu cannot produce this state itself — applying Callout with the figure selected classes both the figure and the image — so paste it in.

1. *View source code* → paste the following, then switch back to rich text:

   ```html
   <figure><img class="callout" src="/umbraco/backoffice/assets/img/installer/logo.svg"><figcaption>caption</figcaption></figure>
   ```

2. Click the image. The whole figure is selected (it is atomic — you cannot place a caret inside it).
3. **Bug:** open Style Select → Containers. Callout is **not** highlighted, though the source shows `class="callout"` on the ``.
4. Click Callout: the class is **removed** from the ``.

Covered by the same fix as Test 4.

**Comment by leekelleher** [comment]
Thanks for the thorough repro steps @AndyButland — really helpful, and all three edge cases checked out exactly as you diagnosed against Tiptap's own `getMarkAttributes`/`getNodeAttributes`.

All one root cause: `isActive` was reading a narrower slice of the document than `execute` acts on. Fixed in 04f2c6d:

- **Test 3** — `#hasMarkWithAttributes` now unions `storedMarks` with `$head.marks()` instead of `storedMarks ?? $from.marks()`, so a mark toggle at the caret (e.g. Ctrl+B) no longer makes `??` discard the marks actually in the document.
- **Tests 4 & 5** — node-attribute detection now walks `nodesBetween(from, to)`, the same range `getNodeAttributes` reads, instead of only the selection's ancestor chain / the selected node's own attrs. This also naturally covers a class on a descendant of a selected atomic node (your figure/image case), so the separate `NodeSelection`-only check was removed as redundant.

Added a test per case (each verified to fail against the previous code first), and manually walked your five repro sequences in the running app — all now behave as expected.

One thing I deliberately left alone: `getNodeAttributes` takes the *last* matching node per type in the range, while our detection returns true on *any* match — so `<p class="callout">first</p><p>second</p>` with both selected reads highlighted, but clicking would add rather than toggle off. That's pre-existing (unchanged by this PR) and you only raised it as a contrast in Test 4 step 6, not as a defect — happy to pick it up separately if you'd like it tightened too, but I didn't want to fold an unrelated invariant change into this fix.

**Review by AndyButland — APPROVED** [review]
Thanks for the updates @leekelleher - looks good now.
